### PR TITLE
Finish user signup

### PR DESCRIPTION
**Pull request by lumosnysm**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,4 +1,26 @@
 class UsersController < ApplicationController
   def new
+    @user = User.new
   end
+
+  def show
+    @user = User.find(params[:id])
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: use find_by instead of find, remove ()

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,4 +1,26 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,26 @@
...
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
+      redirect_to @user
+    else
+      render 'new'
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: use :new instead of 'new'

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,4 +1,26 @@
...
+    end
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password,
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: remove () after permit

File: app/views/shared/_error_messages.html.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, "error") %>.
```
Comment (huuphuc93) [inline app/views/shared/_error_messages.html.erb]: remove ()

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -1,3 +1,24 @@
 <% provide(:title, 'Sign up') %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: remove (), use " instead of '

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -1,3 +1,24 @@
...
 <h1>Sign up</h1>
-<p>This will be a signup page for new users.</p>
\ No newline at end of file
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: signup_path) do |f| %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: remove ()

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -1,3 +1,24 @@
...
+      <%= render 'shared/error_messages' %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: user " instead ' in this whole file

File: config/routes.rb [hunk 9]
```diff
@@ -5,4 +5,6 @@
   get "/about", to: "static_pages#about"
   get "/contact", to: "static_pages#contact"
   get "/signup", to: "users#new"
+  post '/signup',  to: 'users#create'
```
Comment (huuphuc93) [inline config/routes.rb]: use " instead of '

File: app/assets/stylesheets/custom.scss [hunk 10]
```diff
@@ -2,6 +2,14 @@
 @import 'bootstrap';
 @import 'color';
 
+/* mixins, etc. */
+
+@mixin box_sizing {
+  -moz-box-sizing:    border-box;
```
Comment (huongnguyenmta) [inline app/assets/stylesheets/custom.scss]: => `-moz-box-sizing:    border-box;`

File: app/assets/stylesheets/custom.scss [hunk 11]
```diff
@@ -92,4 +100,81 @@ footer {
...
+}
+
+/* miscellaneous */
+
+.debug_dump {
+  clear: both;
+  float: left;
+  width: 100%;
+  margin-top: 45px;
+  -moz-box-sizing:    border-box;
```
Comment (huongnguyenmta) [inline app/assets/stylesheets/custom.scss]: using `mixin`

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -1,4 +1,26 @@
...
   end
+
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: I18n
https://guides.rubyonrails.org/i18n.html

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -1,2 +1,7 @@
 module UsersHelper
+  def gravatar_for(user, size: 80)
```
Comment (huongnguyenmta) [inline app/helpers/users_helper.rb]: using `gem config`, define `5` into settings
https://github.com/railsconfig/config

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -1,3 +1,24 @@
...
+      <%= f.label :name %>
+      <%= f.text_field :name, class: 'form-control' %>
+
+      <%= f.label :email %>
+      <%= f.email_field :email, class: 'form-control' %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
```
Comment (huongnguyenmta) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (huuphuc93) [inline app/views/users/show.html.erb]: remove ()

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -1,4 +1,26 @@
 class UsersController < ApplicationController
   def new
+    @user = User.new
   end
+
+  def show
+    @user = User.find_by params[:id]
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb:7]: find_by id: params[:id], what is different between find_by and find

File: app/views/users/new.html.erb [hunk 17]
```diff
@@ -1,3 +1,24 @@
-<% provide(:title, 'Sign up') %>
+<% provide :title, "Sign up" %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb:1]: I18n

File: app/views/users/new.html.erb [hunk 18]
```diff
@@ -1,3 +1,24 @@
...
+      <%= f.label :name %>
+      <%= f.text_field :name, class: "form-control" %>
+
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.label :password_confirmation, ( t ".confirmation" ) %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb:18]: ( t ".confirmation" ) => t(".confirmation")

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -1,3 +1,24 @@
...
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+
+      <%= f.submit "Create my account", class: "btn btn-primary" %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb:21]: I18n

File: app/views/shared/_error_messages.html.erb [hunk 20]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize @user.errors.count, "error" %>.
```
Comment (huuphuc93) [inline app/views/shared/_error_messages.html.erb:4]: I18n

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (lumosnysm) [reply app/controllers/users_controller.rb:7]: find will return an exception if no record found => server maybe crash, and find_by return nil

**Pull request merged**